feat(output-paths): expose androidOutputPath / iosOutputPath configs

mikehardy opened the pull request:
Fantastic library, thanks for creating this! I used it to generate the default icons for my app,
but I've got multiple flavors/targets in order to push native builds that are wired to different backends,
and I want to have different icons for each (e.g., bright red look out for "Dev", normal for "Prod" etc)

this takes the existing support for android and ios output directories and
exposes it through the config interface with friendly names so that apps with
multiple android variants and multiple ios targets can drop their icons in the
right directories

What do you think?

aeirola commented:
Thanks for the PR! This is a new use case, and seems like something worth supporting. I'll take a closer look at the code a bit later.

aeirola commented on README.md:
```diff
@@ -66,19 +66,23 @@ Behaviour can be configured in the `app.json` under the `svgAppIcon` field. For
+| `backgroundPath`    | `"./icon-background.svg"`                            | Input file path for the background layer. File doesn't need to exist, and will default to a fully white background. If file exist, it needs to be fully opaque. |
+| `platforms`         | `["android", "ios"]`                                 | Array of platforms for which application launcher icons should be generated. Possible values are `android` and `ios`.                                           |
+| `force`             | `false`                                              | When `true`, output files will always be written even if they are newer than the input files.                                                                   |
+| `androidOutputPath` | `./android/app/src/main/res`                         | Where to place generated Android icons                                                                                                                          |
```
Might be good to clarify that this should be the main android resources folder, and that icon files will be placed in the various resource variant specific subfolders.

aeirola commented on README.md:
```diff
@@ -66,19 +66,23 @@ Behaviour can be configured in the `app.json` under the `svgAppIcon` field. For
+| `platforms`         | `["android", "ios"]`                                 | Array of platforms for which application launcher icons should be generated. Possible values are `android` and `ios`.                                           |
+| `force`             | `false`                                              | When `true`, output files will always be written even if they are newer than the input files.                                                                   |
+| `androidOutputPath` | `./android/app/src/main/res`                         | Where to place generated Android icons                                                                                                                          |
+| `iosOutputPath`     | `./ios/*/Images.xcassets/AppIcon.appiconset`         | Where to place generated iOS icons, defaults to first `Images.xcassets` found as a sub-sub-directory in `./ios`                                                 |
```
Might be good to clarify that this should be an `.appiconset` directory. Now the description about the default might confuse someone that it should be an `.xcassets` directory.

aeirola commented on src/cli.ts:
```diff
@@ -61,7 +64,11 @@ async function main(args: string[] = []): Promise<void> {
     platforms: cliConfig.platforms,
-    force: cliConfig.force
+    force: cliConfig.force,
+    androidOutputPath: (await fse.pathExists(cliConfig.androidOutputPath))
+      ? cliConfig.androidOutputPath
+      : undefined,
```
I don't think it makes sense to silently revert to the default if the configured directory doesn't exist. It sounds like it might be confusing for users. We should trust that the user knows what he or she is doing.

aeirola commented on src/cli.ts:
```diff
@@ -61,7 +64,11 @@ async function main(args: string[] = []): Promise<void> {
+    androidOutputPath: (await fse.pathExists(cliConfig.androidOutputPath))
+      ? cliConfig.androidOutputPath
+      : undefined,
+    iosOutputPath: cliConfig.iosOutputPath ?? undefined // no default, determined dynamically in ios.ts
```
Is the `??` actually needed here? Shouldn't the types match already?

aeirola commented on src/cli.ts:
```diff
@@ -75,6 +82,7 @@ async function readFileConfig(): Promise<Partial<CliConfig>> {
     const appJson = (await fse.readJson("./app.json")) as AppJson;
     return appJson.svgAppIcon || {};
   } catch (error) {
+    console.log(`Configuration error: ${JSON.stringify(error)}`);
```
The tool should work without any configurations, so I don't think we should treat this as a configuration error.

aeirola commented on src/cli.ts:
```diff
@@ -112,6 +120,18 @@ async function readArgsConfig(args: string[]): Promise<Partial<CliConfig>> {
+    // --ios-output-path
+    .opt()
+    .name("iosOutputPath")
+    .title("ios-output path")
```
An unintended dash?

aeirola commented on src/cli.ts:
```diff
@@ -112,6 +120,18 @@ async function readArgsConfig(args: string[]): Promise<Partial<CliConfig>> {
+    // --android-output-path
+    .opt()
+    .name("androidOutputPath")
+    .title("android output path")
```
Just a documentation string. Proper capitalisation would be nice.

aeirola reviewed (changes requested):
In general the implementation looks solid. Thanks again for the effort. There were some small inconsistencies that would be good to sort out before merging. If you still have the time and energy to improve the PR, feel free to fix them. Otherwise I can tackle them at some point when I find the time.

mikehardy commented:
Thanks for the review! I'll shape it up and re-push but it will be an "in between things" task so might take a little bit.
All the notes seem spot on.
I'm certain I'll be doing this again (I work almost exclusively on Firebase projects and they are always wired to multiple backends for data separation) so no worries even if this is dormant for a while. I'll actually need to use this for a new project in a couple weeks and that's when I'll be back, using this tool and I'll patch it up  ;-)

aeirola commented:
What is your situation @mikehardy regarding the finalisation of the PR? I'm planning on doing some new releases of the tool, and was hoping to include this feature as well.

If you still think you can find the time to fix the various comments and rebase the changes, I'd be happy to still wait for that. Otherwise, I think I'll just recreate the PR myself and apply the fixes there.

mikehardy commented:
This sat for much longer than intended! Sorry
Yes, I just rebased to current tip of main repo
I added a second commit that contains all the changes requested so they're easy to see, all the comments seemed easy to handle as mentioned months ago :sweat_smile: 
I can squash the two together for one clean commit or you can squash merge or of course if you see anything else on a fresh pass just let me know!

aeirola reviewed (approved):
Thanks for the quick response, and fixes to the comments! The code looks good now, but noticed that the tests don't pass since the changes to the config object haven't been reflected there. Sorry for not running them earlier. Once the tests pass, this will be ready to merge.

mikehardy commented:
I didn't notice there were tests, of course I should have looked and yes the config won't match the fixture now. I'll patch it up

mikehardy commented:
Okay, I pushed a fix for the test fixture generation - sorry I hadn't noticed that before! I understand why Github Actions don't run by default for new contributors but it caught me off guard that there was CI and it just hadn't run :-)

Anyway - maybe worth mentioning that I normally use yarn, which isn't really important except it explains why I noticed trouble in your future you haven't see yet. When you run `npm ci` combined with the `package-lock.json` file here, I can now run `npm test` and it succeeds.

However! If I run `yarn` (equivalent to `npm ci`) there is no lockfile so I get current versions within bounds of version ranges in `.package.json` (equivalent to `npm ci` but without `package-lock.json` existing, in practice). This leads to problems. I was not able to successfully `yarn lint:typescript` or `yarn lint:eslint` this way, for separate reasons, and it might be worth purging your existing package-lock.json (or forcing npm to update all packages) and forward porting to current versions...

aeirola commented:
Thanks for the quick fix once again. I'll merge this and try to get it released in v0.6.0 during this week.

As for the package lockfile. Yeah this is an `npm` project, so probably won't work with yarn. I should probably add some development / contribution documentation in the readme to make it easier to work on. I'll also update the dependencies every once in a while, but that isn't really a reliable solution.

aeirola commented:
Yeah, that is the curse of npm dependencies. Anything seems to be able to break at any update of any dependency at any level.

Personally, I'm not really eager to constantly update everything. Every once in a while, when there is some larger development going on is enough for me.
